src/nrniv/nrnbbcore_write/callbacks/bbcore_callbacks.cpp

```diff
@@ -0,0 +1,524 @@
+#include "bbcore_callbacks.h"
+#include "nrnconf.h"
+#include "nrnmpi.h"
+#include "section.h"
+#include "netcon.h"
+#include "hocdec.h"
+#include "nrnbbcore_write/data/cell_group.h"
+#include "nrnbbcore_write/io/bbcore_write.h"
+#include "parse.h"
+#include "nrnran123.h" // globalindex written to globals.
+#include "netcvode.h" // for nrnbbcore_vecplay_write
+#include "vrecitem.h" // for nrnbbcore_vecplay_write
+
+#if defined(HAVE_DLFCN_H)
+#include <dlfcn.h>
+#endif
+
+extern "C" {
+extern bbcore_write_t* nrn_bbcore_write_;
+extern short* nrn_is_artificial_;
+extern bool corenrn_direct;
+extern int* bbcore_dparam_size;
+extern double nrn_ion_charge(Symbol*);
+extern CellGroup* cellgroups_;
+extern NetCvode* net_cvode_instance;
+extern char* pnt_map;
+
+};
+
+/** Populate function pointers by mapping function pointers for callback */
+void map_coreneuron_callbacks(void* handle) {
+    for (int i=0; cnbs[i].name; ++i) {
+        void* sym = dlsym(handle, cnbs[i].name);
+        if (!sym) {
+            fprintf(stderr, "Could not get symbol %s from CoreNEURON\n", cnbs[i].name);
+            hoc_execerror("dlsym returned NULL", NULL);
+        }
+        void** c = (void**)sym;
+        *c = (void*)(cnbs[i].f);
+    }
+}
+
+void write_memb_mech_types_direct(std::ostream& s) {
+    // list of Memb_func names, types, point type info, is_ion
+    // and data, pdata instance sizes. If the mechanism is an eion type,
+    // the following line is the charge.
+    // Not all Memb_func are necessarily used in the model.
+    s << bbcore_write_version << std::endl;
+    s << n_memb_func << std::endl;
+    for (int type=2; type < n_memb_func; ++type) {
+        const char* w = " ";
+        Memb_func& mf = memb_func[type];
+        s << mf.sym->name << w << type << w
+          << int(pnt_map[type]) << w // the pointtype, 0 means not a POINT_PROCESS
+          << nrn_is_artificial_[type] << w
+          << nrn_is_ion(type) << w
+          << nrn_prop_param_size_[type] << w << bbcore_dparam_size[type] << std::endl;
+
+        if (nrn_is_ion(type)) {
+            s << nrn_ion_charge(mf.sym) << std::endl;
+        }
+    }
+}
+
+
+
+
```

Comment:
nitpick : this would be done by clang-format soon but just to mention - #new lines are inconsistent :)

Reply:
:) that's because of a lot of moving code here an there